lte/gateway/c/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp

```diff
@@ -0,0 +1,247 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <sstream>
+#include <thread>
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include "log.h"
+#include "intertask_interface_types.h"
+#include "intertask_interface.h"
+#include "directoryd.h"
+#include "conversions.h"
+#include "bstrlib.h"
+#ifdef __cplusplus
+};
+#endif
+#include "common_defs.h"
+#include "amf_app_ue_context_and_proc.h"
+#include "ngap_messages_types.h"
+
+using namespace std;
+namespace magma5g {
+extern task_zmq_ctx_t amf_app_task_zmq_ctx;
+/*
+ * AMBR calculation based on 9.11.4.14 of 24-501
+ */
+void ambr_calculation_pdu_session(
+    smf_context_t* smf_context, uint64_t* dl_pdu_ambr, uint64_t* ul_pdu_ambr) {
+  if ((smf_context->dl_ambr_unit == 0) || (smf_context->ul_ambr_unit == 0) ||
+      (smf_context->dl_session_ambr == 0) ||
+      (smf_context->dl_session_ambr == 0)) {
+    // AMBR has not been populated till now and default assigned
+    *dl_pdu_ambr = (64 * 32768);
+    *ul_pdu_ambr = (64 * 32768);
+  } else {
+    // refer 24-501 9.11.4.14
+    switch (smf_context->dl_ambr_unit) {
+      case 1:
+        *dl_pdu_ambr = (1 * smf_context->dl_session_ambr);
+        break;
+      case 2:
+        *dl_pdu_ambr = (4 * smf_context->dl_session_ambr);
+        break;
+      case 3:
+        *dl_pdu_ambr = (16 * smf_context->dl_session_ambr);
+        break;
+      case 4:
+        *dl_pdu_ambr = (64 * smf_context->dl_session_ambr);
+        break;
+        // many more to be coded
+      default:
+        *dl_pdu_ambr = (256 * smf_context->dl_session_ambr);
+        break;
```

Comment:
Isn't this simply (4^(smf_context->dl_ambr_unit-1)) * smf_context->dl_session_ambr for values 1-4 ?

Reply:
Replaced with suggested math equation